lib/Backend/GlobOpt.cpp

```diff
@@ -16142,8 +16142,17 @@ GlobOpt::OptIsInvariant(Sym *sym, BasicBlock *block, Loop *loop, Value *srcVal,
         return false;
     }
 
-    // Can't hoist non-primitives, unless we have safeguards against valueof/tostring.
-    if (!allowNonPrimitives && !srcVal->GetValueInfo()->IsPrimitive() && !loop->landingPad->globOptData.IsTypeSpecialized(sym))
+    // A symbol is invariant if its current value is the same as it was upon entering the loop.
+    loopHeadVal = loop->landingPad->globOptData.FindValue(sym);
+    if (loopHeadVal == NULL || loopHeadVal->GetValueNumber() != srcVal->GetValueNumber())
+    {
+        return false;
+    }
+
+    // Can't hoist non-primitives, unless we have safeguards against valueof/tostring.  Additionally, we need to consider
+    // the value annotations on the source *before* the loop: if we hoist this instruction outside the loop, we can't
+    // necessarily rely on type annotations added (and enforced) earlier in the loop's body.
+    if (!allowNonPrimitives && !loopHeadVal->GetValueInfo()->IsPrimitive() && !loop->landingPad->globOptData.IsTypeSpecialized(sym))
```

Comment:
>loopHeadVal->GetValueInfo()->IsPrimitive() && !loop->landingPad->globOptData.IsTypeSpecialized [](start = 32, length = 94)

@louislaf, I wonder if these two checks check the same thing. Should an int/float type specialized sym have a definitely primitive value info?

Reply:
Yes, you are right.  The IsTypeSpecialized check is probably unnecessary and overly conservative.
Richard, can you try keeping only the IsPrimitive check to see what happens?


---
In reply to: [135135426](https://github.com/Microsoft/ChakraCore/pull/3581#discussion_r135135426) [](ancestors = 135135426)

Reply:
Done; VSO builds both completed with no errors.